docs/designs/checkpoint_design.md

```diff
@@ -0,0 +1,59 @@
+* We can restore the model parameters from a checkpoint to train or predict. Fault tolerance in ElasticDL can not resolve all unexpected error e.g. a power outage and OS fault in the master. We are going to have a bad time if we lose one or more of our experiments due to those errors. Other times, even if we don't experience an unforeseen error, we might just want to resume a particular state of the training for a new experiment or try different things from a given state.
+
+In the following sections, we will describe the design of how to make a checkpoint and restore model parameters from a checkpoint in ElasticDL.
+
+## Design Components
+This design contains two key parts: export and restore.
+
+### Export Model Parameters to a Checkpoint
+
+**When to save a checkpoint during training.** 
+
+In ElasticDL, we will save all model parameters to a checkpoint directory every `checkpoint_steps`. Though, some frameworks support to save checkpoints every N steps or every N seconds, e.g. [Estimator](https://www.tensorflow.org/api_docs/python/tf/estimator/RunConfig#__init__). Using ParameterServerStrategy in ElasticDL, Communication with PS is asynchronous and we cannot guarantee that the iteration steps of all PS instances are consistent at the same time. So, we don't support to save checkpoints every N seconds.
+
+
+**Where to save a checkpoint.** 
+
+We need to set the `checkpoint_dir` in args if we want to save checkpoints. The saved checkpoint directory will be generated by `checkpoint_dir` and iteration steps which we call model version in ElasticDL, e.g. `/{checkpoint_dir}/model_v{version}/`.
+```python
+checkpoint_folder_name = "model_v%s" % version
+checkpoint_dir = os.path.join(
+    args.checkpoint_dir, checkpoint_folder_name
+)
+```
+ElasticDL is a Kubernetes-native framework and local directory in the pod will be lost if the pod exits in Kubernetes. So the `checkpoint_dir` should be a persistent directory which will not be lost after the pod exit. Now, we can utilize `hostPath` and `PersistentVolume` to get a persistent directory to save checkpoints in ElasticDL.
+
+**How to save model parameters to a checkpoint directory.**
+
+Using AllReduceStrategy in ElasticDL, each worker has all parameters and we can assign any worker to save all parameters to a checkpoint directory. Using ParameterServerStrategy, all model parameters are partitioned on all PS instances. They need to save their owner parameters to the same checkpoint directory at the same step. Saving checkpoints for AllReduceStrategy can be seen as a special case that there is only a PS instance using ParameterServerStrategy. So, we can adopt the same way to save checkpoints for AllReduceStrategy and ParameterServerStrategy.
```

Comment:
`they need to save their owner parameters`?  Who is `they`?  
These PS instances need to save their parameter partitions to the same checkpoint directory ...